include/onnxruntime/core/session/onnxruntime_c_api.h

```diff
@@ -1184,6 +1184,16 @@ struct OrtApi {
 */
 #define OrtCustomOpApi OrtApi
 
+// Specifies some characteristics of inputs/outputs of custom ops:
+// Specify if the inputs/outputs are one of:
+// 1) Non-optional (input/output must be present in the node)
+// 2) Optional (input/output may be absent in the node)
+typedef enum OrtCustomOpInputOutputCharacteristic {
+  // TODO: Support 'Variadic' inputs/outputs
+  INPUT_OUTPUT_NONOPTIONAL = 0,
```

Comment:
Design in place to support variadic inputs/outputs in future. This version doesn't support it.

Reply:
nit: INPUT_OUTPUT_REQUIRED?

Reply:
Currently the two states are 'single' or 'optional'. If we add support for variadic does that add 'variadic' as a third option, or we could have 'optional single' and 'optional variadic'? Or would the latter just be a 'variadic' input/output with zero items? A variadic input with zero is supported by ONNX for inputs (e.g. loop state variables for the Loop operator) but not sure if that is allowed for variadic outputs.

Reply:
Fixed the nit. When we do support variadic, we may have variadic_required (needs atleast one) or variadic_optional (can be missing altogether) ? Either way, we can define the enum values when we do get around to supporting variadic.